hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/response/key/OMTrashRecoverResponse.java

```diff
@@ -0,0 +1,64 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ * <p>
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hadoop.ozone.om.response.key;
+
+import org.apache.hadoop.ozone.OmUtils;
+import org.apache.hadoop.ozone.om.OMMetadataManager;
+import org.apache.hadoop.ozone.om.helpers.OmKeyInfo;
+import org.apache.hadoop.ozone.om.helpers.RepeatedOmKeyInfo;
+import org.apache.hadoop.ozone.om.response.OMClientResponse;
+import org.apache.hadoop.ozone.protocol.proto.OzoneManagerProtocolProtos
+    .OMResponse;
+import org.apache.hadoop.hdds.utils.db.BatchOperation;
+
+import java.io.IOException;
+import javax.annotation.Nullable;
+import javax.annotation.Nonnull;
+
+/**
+ * Response for RecoverTrash request.
+ */
+public class OMTrashRecoverResponse extends OMClientResponse {
+  private OmKeyInfo omKeyInfo;
+
+  public OMTrashRecoverResponse(@Nullable OmKeyInfo omKeyInfo,
+      @Nonnull OMResponse omResponse) {
+    super(omResponse);
+    this.omKeyInfo = omKeyInfo;
+  }
+
+  @Override
+  public void addToDBBatch(OMMetadataManager omMetadataManager,
+      BatchOperation batchOperation) throws IOException {
+
+      /* TODO: HDDS-2425. HDDS-2426. */
+    String trashKey = omMetadataManager
+        .getOzoneKey(omKeyInfo.getVolumeName(),
+            omKeyInfo.getBucketName(), omKeyInfo.getKeyName());
+    RepeatedOmKeyInfo repeatedOmKeyInfo = omMetadataManager
+        .getDeletedTable().get(trashKey);
+    omKeyInfo = OmUtils.prepareKeyForRecover(omKeyInfo, repeatedOmKeyInfo);
+    omMetadataManager.getDeletedTable()
+        .deleteWithBatch(batchOperation, omKeyInfo.getKeyName());
+    /* TODO: trashKey should be updated to destinationBucket. */
```

Comment:
One question:
1. if the key is created, deleted, the key is created and the key is deleted. Now when recover, which omKeyInfo will be used from the delete table.

Reply:
Thank you @bharatviswa504 for the question.

Refer to processing of `DeletedTable` in [`OMKeyDeleteResponse#addToDBBatch()`](https://github.com/apache/hadoop-ozone/blob/master/hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/response/key/OMKeyDeleteResponse.java#L82#L88) and [`OmUtils#prepareKeyForDelete()`](https://github.com/apache/hadoop-ozone/blob/master/hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/OmUtils.java#L489#L495), the latest deleted key is added in tail of [`RepeatedOmKeyInfo#omKeyInfoList`](https://github.com/apache/hadoop-ozone/blob/master/hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/helpers/RepeatedOmKeyInfo.java#L35).

So I think we could recover the latest deleted key from the `DeletedTable` in this `created-deleted-created-deleted` situation.

Would you please give me advice if I miss something ?

Reply:
I am fine with recovering last delete key if that is the expected behavior.

>  (And when recovering the latest key, I think we should clear the old deleted key.)
We should not delete the other keys, as those keys will be picked by background trash service and the data for those keys need to be deleted.

And also doing this way, is also not correct from my understanding, let us say, we put those keys in delete table, and background delete key service will pick them up and send to SCM for deletion, at this point we got a recover trash command, so there is a chance that we recover the key which might have no data, as we submitted the request to SCM for deletion, and SCM, in turn, it will send to DN. How we shall handle this kind of scenarios?

Because deletion from delete table will happen when key purge request happens.

Code snippet link [#link]( https://github.com/apache/hadoop-ozone/blob/master/hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/KeyDeletingService.java#L167
)

Reply:
/pending I'm tracing the background part. (Hope soon)

Reply:
Thank you @bharatviswa504 for taking time to review this.

Here is my thought,
We set [`modificationTime`](https://github.com/apache/hadoop-ozone/blob/master/hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/OMKeyDeleteRequest.java#L81L82) when deleting key.

So I think we can compare the `modificationTime` with `RECOVERY_WINDOW` to exclude keys(exist in trash-enabled buckets) from purging.

Code snippet would be added after this [line](https://github.com/apache/hadoop-ozone/blob/master/hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OmMetadataManagerImpl.java#L923) might like 
```Java
if (trashEnable(info.getBucketName()) &&
    (Time.now() - info.getModificationTime()) < RECOVERY_WINDOW) {
    
  /* Would not delete key in this situation. */
}
```
Could you please give me your thoughts or ideas if I miss something, thank you.

And [here](https://issues.apache.org/jira/browse/HDDS-2416?focusedCommentId=17076151&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-17076151) is discussion about trash-recovery.